src/main/java/util/ApiQuotaManager.java

```diff
@@ -0,0 +1,180 @@
+
+        logger.info("Refresh period updated to " + refreshDurationInMinutes
+                + "mins with API calls used in recent refresh cycle is " + apiCallsUsedInRecentRefresh
+                + ", current API quota is " + newRemainingApiQuota + " and next API quota top-up in "
+                + nextRefreshInMinuteFromNow + "mins.");
+    }
+
+    /**
+     * Computes the refreshTimer period that is used for refreshing the issues periodically.
+     * Assumes future refreshes will take the same number of API calls as the last refresh and finds out the refresh
+     * duration that will spread out the refreshes until the next API quota top up.
+     *
+     * @param apiQuota The remaining API quota until the next API quota renewal.
+     *                 Pre-condition: >= 0
+     * @param remainingTimeInMins The remaining time left until the next API quota renewal.
+     *                            Pre-condition: >= 0
+     * @param apiCallsUsedInRecentRefresh The amount of API used in the recent API refresh.
+     *                                      Pre-condition: >= 0
+     * @param apiQuotaBuffer The amount of API calls that is set aside for manual operations by the user.
+     *                       Pre-condition: >= 0
+     * @param minRefreshPeriod The minimal refresh period that will be used.
+     *                         Recommended value : ApiQuotaManager.DEFAULT_REFRESH_PERIOD_IN_MINS
+     *                         Pre-condition: > 0
+     * @return the computed refresh period or minRefreshPeriod.
+     *          minRefreshPeriod will be returned if one of the following conditions are met:
+     *          1) When apiCallsUsedInRecentRefresh == 0 (i.e. when no API quota is used in recent refresh)
+     *          2) When computed refresh period < minRefreshPeriod
+     *                  (i.e. when the remaining API quota is enough to maintain the min refresh rate)
+     *          3) When remainingTimeInMins == 0 (i.e. when a refresh is due to happen inside the next minute).
+     */
+    @SuppressWarnings("PMD") // PMD warning is suppressed to allow explicit parenthesis.
+    public static long computeRefreshTimerPeriod(int apiQuota, long remainingTimeInMins,
+                                                 int apiCallsUsedInRecentRefresh,
+                                                 int apiQuotaBuffer, int minRefreshPeriod) {
+
+        assert apiQuota >= 0 && remainingTimeInMins >= 0 && apiCallsUsedInRecentRefresh >= 0
+                && apiQuotaBuffer >= 0 && minRefreshPeriod > 0;
+
+        int usableApiQuota = apiQuota - apiQuotaBuffer;
+
+        boolean isZeroApiCallsUsedWithEnoughApiQuotaForMinimalRefreshPeriod = apiCallsUsedInRecentRefresh == 0
+                                                                              && usableApiQuota > 0;
+
+        if (isZeroApiCallsUsedWithEnoughApiQuotaForMinimalRefreshPeriod || remainingTimeInMins == 0) {
+            return minRefreshPeriod;
+        }
```

Comment:
These conditions happen from time to time. It's quite common. Since the minutesToNextQuotaTopup is the trimmed value of its millisecs version, there is a 50/50 probability that refresh will occur before the topup.
There is a high chance that apiCalls don't get used between refreshes. Especially there is no usage of update feature and empty board.

I have break these two condition into two different area. As we don't want to return minRefreshPeriod when apiCallsUsedInRecentRefresh == 0 and there is no usableApiQuota


```
        if (minutesToNextQuotaTopup == 0) {
            return minRefreshPeriod;
        }

        long refreshTimeInMins;

        if (usableApiQuota < apiCallsUsedInRecentRefresh) {
            refreshTimeInMins = minutesToNextQuotaTopup;
            return Math.max(refreshTimeInMins, minRefreshPeriod);
        }

        if (apiCallsUsedInRecentRefresh == 0){
            return minRefreshPeriod;
        }
```